processing/vector-indexes.mdx

```diff
@@ -0,0 +1,185 @@
+
+## Vector distance operators
+
+RisingWave provides specialized operators for calculating vector distances:
+
+| Operator | Function | Description |
+| :--- | :--- | :--- |
+| `<->` | `l2_distance()` | Euclidean (L2) distance |
+| `<=>` | `cosine_distance()` | Cosine distance |
+| `<+>` | `l1_distance()` | Manhattan (L1) distance |
+| `<#>` | Negative inner product | Negative inner product distance |
+
+## Vector similarity search
+
+Use vector distance operators with `ORDER BY` and `LIMIT` to perform similarity search:
+
+```sql
+-- Find the 5 most similar items using L2 distance
+SELECT * FROM items 
+ORDER BY embedding <-> '[3,1,2]' 
+LIMIT 5;
+
+-- Find similar items using cosine distance
+SELECT id, name FROM items 
+ORDER BY embedding <=> '[0.5, 0.3, 0.2]' 
+LIMIT 10;
+```
+
+## Vector data type, operators, functions
+
+RisingWave provides a `vector(n)` data type where `n` specifies the dimension of the vector.
+
+```sql
+-- Create a table with vector columns
+CREATE TABLE items (
+    id INT PRIMARY KEY,
+    name STRING,
+    embedding vector(3)
+) APPEND ONLY;
+```
+
+RisingWave also provides various operators and functions for working with vectors. See [Vector data type](/sql/data-types/vector) for details.
+
+## Vector indexes on function expressions
+
+You can create vector indexes on function expressions instead of raw columns. This allows you to avoid storing a separate vector column, saving storage and reducing maintenance costs.
+
+1. Create or alter the table to include the input column
+
+```sql
+ALTER TABLE items ADD COLUMN description STRING;
+```
+
+2. Define the user-defined function (UDF)
+
+```sql
+CREATE FUNCTION get_embedding(string) RETURNS VECTOR(128) LANGUAGE SQL AS $$
+SELECT openai_embedding('{"model": <EMBEDDING_MODEL_NAME>, "api_key": <API_KEY>}'::jsonb, $1)::vector(128);
+$$;
+```
+
+3. Create the vector index on the function expression
+
+```sql
+CREATE INDEX idx_embedding_func ON items
+USING FLAT (get_embedding(description))
+INCLUDE(description)
+WITH (distance_type = 'l2');
+```
+
+In this example, `get_embedding(description)` is used as the index expression.
+
+This approach avoids materializing a separate vector column in the table, which reduces storage costs and keeps the table schema simpler.
+
+## Examples
+
+### Basic vector similarity search
+
+```sql
+-- Create table with vector data
+CREATE TABLE products (
+    id INT PRIMARY KEY,
+    name STRING,
+    embedding vector(128)
+) APPEND ONLY;
+
+-- Create vector index
+CREATE INDEX idx_embedding ON products 
+USING HNSW (embedding) 
+WITH (distance_type = 'cosine');
+
+-- Insert sample data
+INSERT INTO products VALUES 
+(1, 'Product A', '[0.1, 0.2, ...]'),
+(2, 'Product B', '[0.3, 0.4, ...]');
+
+-- Find similar products
+SELECT id, name 
+FROM products 
+ORDER BY embedding <=> '[0.2, 0.3, ...]' 
+LIMIT 5;
+```
+
+### Using cosine distance type
+
+```sql
```

Comment:
The SQL query should depend on which kind of index we created.

If we create vector index from raw vector column (`USING HNSW (embedding)`), we query with raw column `order by embedding <=> '[0.5, 0.5, 0.0]'`. If created with function expression, we query with function expression `order by get_embedding(description) <=> '[1.0, 2.0, 3.0]'`.